### ISSUE TITLE
Introduce support for nested properties

### ISSUE THREAD

**Issue by jdegoes** [zio/zio-config#68]
A `Map[String, String]` as the basis for properties means that no nested configuration can be supported, which is inherently anti-modular; in addition, many third-party libraries that could work well with ZIO Config support nesting (HOCON, YAML, JSON, etc.).

To support nesting, it would be necessary to transition away from `Map[String, String]` to an alternate data structure. This data structure would _support_ nesting but not require it: in the sense that it should be possible to create a value of this type from any `Map[String, String]`, so that flat formats are supported. Similarly, it should be possible (probably?) to "flatten" the type into something completely flat, so we can read even nested structures from flat sources like environment variables.

If we take into consideration the design discussion around `Pure`, then we will end up with a design similar to the following:

```scala
sealed trait PropertyTree
object PropertyTree {
  final case class Leaf(value: String) extends PropertyTree
  final case class Record(value: Map[String, PropertyTree]) extends PropertyTree

  def fromMap(map: Map[String, String]): PropertyTree = ???
}
```

Probably, we should make that polymorphic and defer specialization to the point where we need it:

```scala
sealed trait PropertyTree[+K, +V] {
  def flatten(zero: K, append: (K, K) => K): Map[K, V] = ???
}
object PropertyTree {
  final case class Leaf[V](value: V) extends PropertyTree[Nothing, V]
  final case class Record[K, V](value: Map[K, PropertyTree[K, V]]) extends PropertyTree[K, V]

  def fromMap(map: Map[String, String]): PropertyTree[String, String] = ???

  def unflatten[K, V](map: Map[List[K], V]): PropertyTree[K, V] = ???
}
```

Using a structure like this, it should be possible to support nested configuration formats.

Note that currently we will have in all cases `PropertyTree[String, String]`. That's because our `PropertyType` is a string-based codec. We could think about the connection there and whether to break the dependency on string or not, but moving away from strings would definitely require a change to `PropertyType`.

**Comment by afsalthaj** [comment]
@jdegoes 
When it comes to nested structure, should write back functionality be considering it writing back to nested structure ? 

On your append and zero usage to flatten it back to map, does that consider when the value is list?

**Comment by afsalthaj** [comment]
This looks really interesting, and hopefully we can have a PR soon to discuss further.

**Comment by afsalthaj** [comment]
Hi John?
would you mind helping us on  integrating it with `ConfigSource`?

**Comment by jdegoes** [comment]
The simplest possible modification to `ConfigSource`:

```scala
def getConfigValue(path: String): IO[ReadError, PropertyTree[String, String]]
```

Runner-up:

```scala
def getConfigValue(paths: List[String]): IO[ReadError, String]
```

Both have enough structure to read nested configs. The latter is a little more "elemental".

We'll have to modify some other places to support nesting when defining properties, etc.

**Comment by afsalthaj** [comment]
The only immediate issue that I could find is:
Getting a map of entire system env may go through security management exception.
while system.env(path) returning Option may still be successful.

This means forming a sys env to Property[String, string] (or even map[String, String]) can be a bit of trouble?

As of now we use zio.system.env and that works well. it either give a ReadError which can be fatal, or missing or a single value, given a key.

**Comment by afsalthaj** [comment]
Or is it that, an instance of `PropertyTree[String, String]` is valid only for 1 path?

**Comment by afsalthaj** [comment]
What could be a proper interface to talk about nested strucuture from a user perspective.

```
    case class Prod(auth: String, port: Int, dbUrl: Option[String])
 
  val config  
     nested("auth") >~> string("ldap") |@| nested("database") >~> int("PORT") |@|
      nested("database") >~> string("DB_URL").optional)(Prod.apply, Prod.unapply)
```

val result = read(config).provide(hocconSource) === Prod(...)
write(config, result) == 
```scala
List(
   PropertyTree.Record(
      "auth", 
       Map("ldap" -> PropertyType.Leaf("sample1")),
   PropertyTree.Record(
     "database",
      Map("PORT" -> PropertyType.Leaf("sample2"), "DB_URL" -> PropertyType.Lead("sample3")
   )
)
```
This allows user to write a valid hoccon, that could be read by some other application expecting the hoccon source.

**Comment by afsalthaj** [comment]
Users can also choose to have 
  
```scala
  string(List("auth", "ldap")) |@| int(List("database", "port")) |@| string(List("database", "DB_URL")).optional
```

In this case, Writing it back to a PropertyTree (hoccon, for instance) can turn out to be a bit fragile?

**Comment by afsalthaj** [comment]
@jdegoes The way we see it as, the description of the config `ConfigDescriptor` should say that, hey, this is a nested config u r expecting - probably it is a hoccon, json, or yaml source?

**Comment by afsalthaj** [comment]
In both approaches, it will be `PropertyTree` that's our initial structure.

**Comment by jdegoes** [comment]
You need something like:

```scala
nested("foo") { /* inside here is a ConfigDescriptor */ } |@| ...
```

The type will be:

```scala
def nested[A](path: String)(cd: ConfigDescriptor[A]): ConfigDescriptor[A]
```

It will require a new case in the `ConfigDescriptor` enumeration.

**Comment by jdegoes** [comment]
By the way, we can do nested even with flat properties.

I used to see this format all the time with Java properties:

```
application.settings.connectionUrl = ...
application.settings.backupUrl = ...
```

With environmental variables you sometimes see `JAVA_ENV`, `JAVA_HOME`, etc.

Because `ConfigDescriptor` is just data, we can translate the nesting into any format we want.

**Comment by jdegoes** [comment]
> The only immediate issue that I could find is:
> Getting a map of entire system env may go through security management exception.
> while system.env(path) returning Option may still be successful.

This implies `ConfigSource` should not support enumeration, e.g. be like:

```scala
def getConfigValue(paths: List[String]): IO[ReadError, String]
```

**Comment by afsalthaj** [comment]
Sure. In any case, it seems we need a new node `Nested` in `ConfigDescriptor`.

**Comment by jdegoes** [comment]
@afsalthaj Yes, that's a great & safe start, since we know we'll need it and what the type should be.

**Comment by afsalthaj** [comment]
Exactly. And what I can see right now before I start development,
Whenever program encounters Nested, it accumulates the `path` which is then used to go to deep into the `PropertyTree`, to fetch the `final` value.

Will need to start scratching with the code now.

**Comment by afsalthaj** [comment]
Fixed in #78
